Attaching api for login

smaliszewski15 opened the pull request:
Api is mostly there for logging in. Also worked on the user profile page so that i could test whether the data was received from the server or not.

smaliszewski15 committed:
Added bottom buttons to all pages

smaliszewski15 committed:
Attached api mostly to sign in

smaliszewski15 committed:
Attached api mostly to sign in

smaliszewski15 committed:
Sign in fixed, other changes made:
User profile page pretty much done
User profile page connected to api
All profile page buttons exist and are usable

nashirj commented on lib/utils/globals.dart:
```diff
@@ -0,0 +1,11 @@
+}
+
+const String API_PREFIX = "https://api-smart-chef.herokuapp.com/";
+const Color greenScheme = Color(0xff008600);
```
Newline at end of file please, ditto elsewhere

nashirj replied:
Instead of having one file for all your misc global stuff, can you make a `colors.dart` for all your color stuff?

nashirj commented on lib/utils/getAPI.dart:
```diff
@@ -1,29 +1,107 @@
+  static late String refreshToken;
+  static late int lastSeen;
+  //static late bool hasProfileImage;
+  //static late String profileImage;
```
Please add a TODO for commented out lines with an associated issue number or delete the commented out lines

nashirj commented on lib/utils/getAPI.dart:
```diff
@@ -1,29 +1,107 @@
+  static late String password;
+  static late String accessToken;
+  static late String refreshToken;
+  static late int lastSeen;
```
Is there a way around using `static late`? Could we create a singleton instance of the user class and fill out all the values? I skimmed [this article](https://dart.dev/null-safety/understanding-null-safety#late-variables) and it looks like late variables are checked everytime they are accessed to ensure they are non-null which may slow down our app. If not, it's ok - just thought I'd ask

smaliszewski15 replied:
i had read that singletons are not really the greatest to use, but i did change it so that late isnt used for every variable, just one. it should hopefully cause less computations to be made for checking nullity.

nashirj commented on lib/utils/getAPI.dart:
```diff
@@ -1,29 +1,107 @@
+      response = await http.put(Uri.parse(totalurl), headers: headers);
+      return response;
     }
     catch(e) {
```
Check out the [dart style guide](https://dart.dev/guides/language/effective-dart/style#do-use-curly-braces-for-all-flow-control-structures) for conventions; it seems like they place the start of the next conditional block on the same line as the brace, e.g.:
```
try {
  // stuff...
} catch(e) {
  // other stuff
}
```
and same elsewhere for if/else statements, etc. You can probably set up your editor to automatically fix this instead of doing it manually

nashirj replied:
Looks like there is a [`dart format`](https://dart.dev/guides/language/effective-dart/style#do-format-your-code-using-dart-format) command that you can use

nashirj commented on lib/utils/getAPI.dart:
```diff
@@ -1,29 +1,107 @@
-    String ret = "";
+import 'package:smart_chef/utils/globals.dart';
+
+class UserData {
```
Is there a reason for using a class to wrap these requests? I think each function can be free floating, especially since you don't use `this` at all.

smaliszewski15 replied:
with the change in place that each endpoint will have its own file, it makes more sense, to me at least, that all the functions are wrapped in their respective class files.
Also, when I tried to have them free floating, the ide was yelling at me that the command isnt recognized. i might have just done something wrong tho, like forget to import the file when i started doing it.

nashirj commented on lib/screens/StartupScreen.dart:
```diff
@@ -839,20 +691,483 @@ class _StartPageState extends State<StartPage> {
+                                String firstName = _firstName.value.text.trim();
+                                String lastName = _lastName.value.text.trim();
+
+                                try {
+                                  String payload = '{"firstname": "$firstName","lastname": "$lastName","username": "$username","password": "$password"}';
+                                  final ret = await UserData.authUser(
+                                      'auth/register', payload);
+                                  switch (ret.statusCode) {
+                                    case 200:
+                                      errorMessage = '';
+                                      _email.clear();
+                                      _password.clear();
+                                      _firstName.clear();
+                                      _lastName.clear();
+                                      setState(() {
+                                        state = 1;
+                                      });
+                                      break;
+                                    case 400:
+                                      print("Incorrect formatting!");
+                                      break;
+                                    case 401:
+                                      errorMessage =
+                                      'Access token is missing or invalid';
+                                      break;
+                                    case 404:
+                                      errorMessage =
+                                      'Email/password incorrect';
+                                      break;
+                                    default:
+                                      errorMessage = 'Something went wrong!';
+                                      break;
+                                  }
+
+                                  //jsonObject = json.decode(ret.body);
+                                }
+                                catch (e) {
+                                  print('Could not connect to server');
+                                }
+                              }
```
This is a lot of stuff happening here that I think could be extracted into functions:
1) validate form (can be done in a helper function in this file)
2) parsing request (can make a new util file called `middleware.dart` that parses/extracts info from API response)

This would improve readability, but in case you need to reuse any of that logic later (which for other types of requests is more likely), it'll be easier to do so.

nashirj replied:
Ditto everywhere else you have this structure where all the request validation/parsing is handled directly in the view

nashirj commented on lib/screens/StartupScreen.dart:
```diff
@@ -839,20 +691,483 @@ class _StartPageState extends State<StartPage> {
+                                      break;
+                                  }
+
+                                  //jsonObject = json.decode(ret.body);
```
Please delete this commented out line

nashirj commented on lib/screens/StartupScreen.dart:
```diff
@@ -839,20 +691,483 @@ class _StartPageState extends State<StartPage> {
+                                  setState(() => unfilledEmail = true);
+                                }
+                              } else {
+                                //var jsonObject;
```
Pls delete

nashirj commented on lib/utils/getAPI.dart:
```diff
@@ -1,29 +1,107 @@
+import 'package:smart_chef/utils/globals.dart';
+
+class UserData {
+  static Future<http.Response> authUser(String url, String outgoing) async{
```
Please add a space between the bracket at the end of `async{` and similar below

nashirj commented on lib/utils/getAPI.dart:
```diff
@@ -1,29 +1,107 @@
+    final headers = {HttpHeaders.contentTypeHeader: 'application/json', HttpHeaders.authorizationHeader: LocalData.accessToken};
+    try {
+      String totalurl = '$API_PREFIX$url';
+      response = await http.put(Uri.parse(totalurl), headers: headers);
```
Delete should use `http.delete`

nashirj commented on lib/screens/StartupScreen.dart:
```diff
@@ -39,678 +42,514 @@ class _StartPageState extends State<StartPage> {
+                                    errorMessage = 'Could not connect to server';
+                                    print('Could not connect to /auth/user');
+                                  }
+                                  /*LocalData.firstName = 'Joe';
+                                  LocalData.lastName = 'Mama';
```
:laughing:

nashirj commented on lib/utils/getAPI.dart:
```diff
@@ -1,29 +1,107 @@
+  static Future<http.Response> logoutUser(String url, String auth) async{
+    http.Response response;
+    final headers = {HttpHeaders.contentTypeHeader: 'application/json', HttpHeaders.authorizationHeader: auth};
+    try {
+      String totalurl = '$API_PREFIX$url';
+      response = await http.get(Uri.parse(totalurl), headers: headers);
+      return response;
+    }
+    catch(e) {
+      print(e.toString());
+    }
+    throw Exception('Could not connect to server');
```
This is identical for most of your requests; could you extract one of these for each of GET, PUT, POST, and DELETE?

nashirj commented on lib/utils/getAPI.dart:
```diff
@@ -1,29 +1,107 @@
-/*import 'package:http/http.dart' as http;
+import 'dart:io';
```
The name of this file should be updated since you have multiple types of requests. You can keep all the requests in one file, but personally I think it would make sense to break up the requests into category of request per file:
- userAPI
- foodAPI
- ...

One per main section of the [backend API](https://app.swaggerhub.com/apis/mplekunov/Smart-Chef/1.0.0#/)

And then you can have an APIutil.dart file that has the base request types, the `API_PREFIX` var, everything you need to use in the downstream API files.

smaliszewski15 committed:
Made most requested changes

smaliszewski15 committed:
Added framework for displaying ingredients on ingredients page

smaliszewski15 committed:
Fixed problems with login

smaliszewski15 committed:
Fixed problems with login

nashirj commented on lib/utils/userData.dart:
```diff
@@ -0,0 +1,24 @@
+  int lastSeen;
+  //TODO: Add profile image support (issue #37)
+  //static late bool hasProfileImage;
+  //static late String profileImage;
```
Pls put a space between the `//` and the start of the comment e.g. `// TODO`. Also use the syntax `// TODO(37): description` so that when you're working on issue 37 you can search the code for `TODO(37)`

nashirj commented on lib/utils/userAPI.dart:
```diff
@@ -0,0 +1,57 @@
+import 'package:smart_chef/utils/APIutils.dart';
+
+class User {
+  static const String extension = 'user';
```
`extension` is a reserved identifier; it's possible to redefine it, but let's not. Can you use something like `endpoint` or `route` instead? Ditto elsewhere that you use `extension` as a variable name. P.S. you can tell if you're using a reserved word if it is red colored in the GitHub UI.

nashirj commented on lib/utils/userAPI.dart:
```diff
@@ -0,0 +1,57 @@
+    http.Response response;
+    final headers = {
+      HttpHeaders.contentTypeHeader: 'application/json',
+      HttpHeaders.authorizationHeader: user.refreshToken
```
I believe you need to use `user.accessToken` to authenticate.

nashirj commented on lib/utils/userAPI.dart:
```diff
@@ -0,0 +1,57 @@
+class User {
+  static const String extension = 'user';
+
+  static Future<http.Response> getUser() async {
```
Please add a newline between main sections of your functions; e.g. in this one after making your variables, and one after the try catch block to make it more readable. Ditto elsewhere

nashirj commented on lib/utils/userAPI.dart:
```diff
@@ -0,0 +1,57 @@
+    };
+    try {
+      String totalurl = '$API_PREFIX$extension';
+      response = await http.get(Uri.parse(totalurl), headers: headers);
```
Since you only use `totalurl` once, you can construct it inline:
```dart
response = await http.get(Uri.parse('$API_PREFIX$extension'), headers: headers);
```

In this case, it doesn't make the line too long or hurt readability; use your judgement as to where else you can condense like this

nashirj commented on lib/utils/authAPI.dart:
```diff
@@ -0,0 +1,71 @@
+          encoding: Encoding.getByName("utf-8"));
+      return response;
+    } catch (e) {
+      print("Could not connect!");
```
Other places you print `e.toString()`, pls have that everywhere so that it's easier for you to debug errors if something goes wrong

nashirj commented on lib/routes/routes.dart:
```diff
@@ -3,22 +3,34 @@ import 'package:smart_chef/screens/StartupScreen.dart';
+    editUserProfileScreen: (context) => EditUserProfilePage(),
+    editPasswordScreen: (context) => EditPasswordPage(),
   };
 }
```
Newline at end of file

nashirj commented on lib/screens/StartupScreen.dart:
```diff
@@ -39,678 +45,460 @@ class _StartPageState extends State<StartPage> {
+                                String password = _password.value.text.trim();
+
+                                try {
+                                  String payload = '{"username": "$username","password": "$password"}';
```
Instead of passing a string like this, I think it would be cleaner to use build in Dart types until you are about to send the payload. This [page](https://docs.flutter.dev/development/data-and-backend/json) talks about how to handle JSON the idiomatic Dart way. tl;dr, can you try to do something like (what I sent is probably not the correct syntax)
```dart
Map<String, dynamic> userPayload = {
  username: username,
  password: password,
};
```
and then only when you're about to send the payload, do you do `json.encode(payload)`?

If you want to implement the stuff from the link I sent and make everything JSON serializable, you are welcome to do that, but I definitely don't expect you to.

nashirj commented on lib/utils/authAPI.dart:
```diff
@@ -0,0 +1,71 @@
+class Authentication {
+  static const String extension = 'auth';
+
+  static Future<http.Response> login(String outgoing) async {
```
See my comment about passing a map instead of a string and consider updating all the parameters of your API requests to use strong types instead of `String`

nashirj commented on lib/utils/authAPI.dart:
```diff
@@ -0,0 +1,71 @@
+    throw Exception('Could not connect to server');
+  }
+
+  static Future<http.Response> refreshToken() async {
```
If refreshing the token fails, you should attempt to log the user in again using their stored `username` and `password`, returning an error if you are unable to do so -> if that happens, you should redirect the user to the login page (from the place where this code is called)

nashirj commented on lib/utils/authAPI.dart:
```diff
@@ -0,0 +1,71 @@
+
+  static Future<http.Response> login(String outgoing) async {
+    http.Response response;
+    final headers = {HttpHeaders.contentTypeHeader: 'application/json'};
```
It looks like this is the same in all of the functions; can you instead define this in APIutils.dart?

nashirj commented on lib/utils/authAPI.dart:
```diff
@@ -0,0 +1,71 @@
+      response = await http.post(Uri.parse(totalurl),
+          body: utf8.encode(outgoing),
+          headers: headers,
+          encoding: Encoding.getByName("utf-8"));
```
What does this do? I see you have it some places and not others.

smaliszewski15 replied:
it was from the professors code. It is used to encode the payload string into json. if i do end up changing the way that the function receives the paylod, then that probably won't end up staying there. It is only really used in any request that will have a request body in it, which is why it isnt in every function

nashirj commented on lib/utils/authAPI.dart:
```diff
@@ -0,0 +1,71 @@
+    } catch (e) {
+      print(e.toString());
+    }
+    throw Exception('Could not connect to server');
```
The way you have this set up is fine, but one common alternate way I see often is:
```dart
    http.Response response;
    final headers = {
      HttpHeaders.contentTypeHeader: 'application/json',
      HttpHeaders.authorizationHeader: user.accessToken
    };

    try {
      String totalurl = '$API_PREFIX$extension/logout';
      response = await http.get(Uri.parse(totalurl), headers: headers);
    } catch (e) {
      print(e.toString());
      throw Exception('Could not connect to server');
    }

    return response;
```

The advantage to this is that you can see without looking through the whole function what the return value is at the end.

If you keep it the way it is (which is fine), you can define response inside the `try` block since it's not used outside:
```dart
http.Response response = await http.get(Uri.parse('$API_PREFIX$extension/logout'),
                                        headers: headers);
return response;
```

nashirj commented on lib/routes/routes.dart:
```diff
@@ -3,22 +3,34 @@ import 'package:smart_chef/screens/StartupScreen.dart';
 
 class Routes {
   static const String startupScreen = '/startup';
+  static const String signinScreen = '/signin';
```
Pls be consistent with `login` vs `signin` (screen is named signin but endpoint is login)

nashirj commented on lib/screens/IngredientScreen.dart:
```diff
@@ -57,43 +59,140 @@ class _IngredientsPageState extends State<IngredientsPage> {
+                        IconButton(
+                          onPressed: () {},
+                          icon: Icon(Icons.egg),
+                          iconSize: 55,
```
Everywhere you have numbers like this that are hardcoded and used in more than one place, please add those to a util file so that it's easy to update them all at one time.

nashirj commented on lib/screens/IngredientScreen.dart:
```diff
@@ -57,43 +59,140 @@ class _IngredientsPageState extends State<IngredientsPage> {
+      child:
+    );
+  }
+}*/
```
I'm not sure if your editor supports this but commonly you can do `ctrl + /` to comment/uncomment multiple lines at once; if that's supported in your editor, can you please do that instead of using `/* */`?

smaliszewski15 replied:
didnt know that existed tbh

nashirj commented on lib/screens/RecipeScreen.dart:
```diff
@@ -30,6 +34,148 @@ class _RecipePageState extends State<RecipePage> {
+              Icons.search,
+              color: Colors.black,
+            ),
+            iconSize: 35,
```
I notice `iconSize` has a different value on this screen; having these defined somewhere (as mentioned in my other comment) will make the UI more consistent

smaliszewski15 replied:
This is a different size because it is in a different location as oppose to other icons. This one is in the top appbar, whereas the others are in the bottom appbar. It wouldn't fit if it were bigger than it currently is.

nashirj commented on lib/screens/RecipeScreen.dart:
```diff
@@ -30,6 +34,148 @@ class _RecipePageState extends State<RecipePage> {
+          centerTitle: true,
+          backgroundColor: Colors.white,
+          leading: IconButton(
+            onPressed: () {},
```
If these `onPressed` actions should do eventually do something, please add corresponding TODOs and issue numbers so that it's easy for you to come back and fix later, ditto elsewhere

nashirj commented on lib/screens/RecipeScreen.dart:
```diff
@@ -30,6 +34,148 @@ class _RecipePageState extends State<RecipePage> {
+            width: MediaQuery.of(context).size.width,
+            height: MediaQuery.of(context).size.height,
+            decoration: BoxDecoration(color: Colors.white),
+            child: Text("To be changed"),
```
Add TODO and issue num for stuff that is WIP

nashirj commented on lib/screens/StartupScreen.dart:
```diff
@@ -39,678 +45,460 @@ class _StartPageState extends State<StartPage> {
+    fillColor: const Color(0xffD1D1D1),
+    border: OutlineInputBorder(
+      borderRadius: BorderRadius.circular(10),
+      borderSide: const BorderSide(color: Color(0xff47A1E2)),
```
If these colors are used elsewhere, feel free to add them to colors.dart

nashirj commented on lib/screens/StartupScreen.dart:
```diff
@@ -39,678 +45,460 @@ class _StartPageState extends State<StartPage> {
+                                        color: Colors.red),
+                                      hintText: 'Enter Email')
+                                : generalDecoration.copyWith(hintText: 'Enter Email'),
+                            onChanged: (password) {
+                              setState(() => unfilledPassword = false);
```
If someone enters a password then deletes it all, will this handle that correctly?

smaliszewski15 replied:
i added support for this in other places, just forgot to add it here.

nashirj commented on lib/screens/StartupScreen.dart:
```diff
@@ -39,678 +45,460 @@ class _StartPageState extends State<StartPage> {
+                                      accessToken: data["accessToken"],
+                                      refreshToken: data["refreshToken"],
+                                      password: password,
+                                    );
```
An advantage to implementing the JSON serializable stuff I mentioned in the other comment is that you would be able to do something like `user = UserData.fromJson(ret.body)`, which would make this a lot cleaner!

nashirj commented on lib/screens/StartupScreen.dart:
```diff
@@ -840,19 +635,490 @@ class _StartPageState extends State<StartPage> {
+        print('Something in auth went wrong!');
+        return false;
+    }
+  }
```
Consider naming this `getErrorString` and have it return an error string. Then at the calling function you can have:
```dart
if (res.statusCode == 200) {
  // success stuff
} else {
  errorMessage = getErrorString(res.statusCode);
  // error stuff, like set error message/render stuff that went wrong
}
```
I wasn't sure exactly what `status` did until looking at this; it's nice to minimize the need to jump to other functions when reading other sections of code - not only for other readers but also for future you :)

nashirj commented on lib/screens/StartupScreen.dart:
```diff
@@ -39,678 +45,460 @@ class _StartPageState extends State<StartPage> {
+                                String username = _email.value.text.trim();
+                                String password = _password.value.text.trim();
+
+                                try {
```
This is much cleaner now after your refactoring, great work!

nashirj commented on lib/screens/StartupScreen.dart:
```diff
@@ -785,46 +580,46 @@ class _StartPageState extends State<StartPage> {
-                    introMessage = 'Welcome To\nSmartChef!';
-                  }
+                if (areFieldsValid(false)) {
+                  //TODO Add support for reseting password
```
Pls add a corresponding issue number

nashirj commented on lib/screens/StartupScreen.dart:
```diff
@@ -785,46 +580,46 @@ class _StartPageState extends State<StartPage> {
-                    });
-                    introMessage = 'Welcome To\nSmartChef!';
-                  }
+                if (areFieldsValid(false)) {
```
I'm not sure if Dart supports it, but can you experiment with trying to match [this style recommendation](https://google.github.io/styleguide/cppguide.html#Function_Argument_Comments) for function argument comments?
```dart
if (areFieldsValid(/*hasPassword=*/false) {
...
}
```

nashirj commented on lib/screens/StartupScreen.dart:
```diff
@@ -840,19 +635,490 @@ class _StartPageState extends State<StartPage> {
-  final _code = TextEditingController();
-  final codeDecoration = InputDecoration(
+  bool areFieldsValid(bool hasPassword) {
+    bool toReturn = true;
```
`allFieldsValid`

nashirj commented on lib/screens/StartupScreen.dart:
```diff
@@ -840,19 +635,490 @@ class _StartPageState extends State<StartPage> {
+    if (hasPassword & _password.value.text.isEmpty) {
+      toReturn = false;
+      setState(() => unfilledPassword = true);
+    }
```
Since you set an error message for invalid emails, wdyt about setting one for the other cases as well?

nashirj commented on lib/screens/StartupScreen.dart:
```diff
@@ -840,19 +635,490 @@ class _StartPageState extends State<StartPage> {
 
-  final _code = TextEditingController();
-  final codeDecoration = InputDecoration(
+  bool areFieldsValid(bool hasPassword) {
```
You have two `areFieldsValid` that are differentiated only by the parameter list, but it seems like they're used for diff purposes - this is for login, and the other is for register. Can you update the function names accordingly? `areLoginFieldsValid` and `areRegisterFieldsValid`?

nashirj commented on lib/screens/StartupScreen.dart:
```diff
@@ -840,19 +635,490 @@ class _StartPageState extends State<StartPage> {
+                              if (!isEmail(email)) {
+                                errorMessage = 'Email must be in valid form';
+                                setState(() => unfilledEmail = true);
+                              }
+                              else {
```
These should be on the same line `} else {`. Ditto elsewhere in places where you have try catch blocks etc. Should be able to use a formatter to fix

nashirj commented on lib/screens/StartupScreen.dart:
```diff
@@ -840,19 +635,490 @@ class _StartPageState extends State<StartPage> {
+                                String lastName = _lastName.value.text.trim();
+
+                                try {
+
```
Don't need newline at start

nashirj commented on lib/screens/StartupScreen.dart:
```diff
@@ -840,19 +635,490 @@ class _StartPageState extends State<StartPage> {
+                                  String payload = '{"firstName": "$firstName","lastName": "$lastName","username": "$username","password": "$password"}';
+                                  final ret = await Authentication.register(payload);
+                                  if (status(ret)) {
+                                    var data = json.decode(ret.body);
+                                    user = UserData(
+                                      firstName: data['userInfo']['firstName'],
+                                      lastName: data['userInfo']['lastName'],
+                                      email: data['userInfo']['username'],
+                                      lastSeen: data['userInfo']['lastSeen'],
+                                      accessToken: data["accessToken"],
+                                      refreshToken: data["refreshToken"],
+                                      password: password,
+                                    );
+
+                                      errorMessage = '';
+                                      clearFields();
+                                      //TODO Add support for verifying registered email
+                                      /*setState(() {
+                                        state = 1;
+                                      });*/
+                                      Navigator.pushNamedAndRemoveUntil(context,
+                                          '/food', ((Route<dynamic> route) => false));
```
The alignment is messed up here, there are different indentation levels in the same block

nashirj commented on lib/screens/StartupScreen.dart:
```diff
@@ -840,19 +635,490 @@ class _StartPageState extends State<StartPage> {
+                        children: <Widget>[
+                          Flexible(
+                            child: Text(
+                                'Already a user? Click here to sign in instead'),
```
"Already have an account?"

nashirj commented on lib/screens/StartupScreen.dart:
```diff
@@ -946,19 +1214,19 @@ class _StartPageState extends State<StartPage> {
                 } else {
+                  _code.clear();
                   setState(() {
-                    _code.clear();
-                    state = 0;
+                    Navigator.pushNamedAndRemoveUntil(
+                        context, '/food', (Route<dynamic> route) => false);
```
Looks like this doesn't actually verify the code - can you add a TODO to fix that later or update this to actually use the code?

nashirj commented on lib/screens/StartupScreen.dart:
```diff
@@ -973,4 +1241,60 @@ class _StartPageState extends State<StartPage> {
+    _lastName.clear();
+  }
+
+  bool status(http.Response res) {
```
You have this defined twice in the same file with the same name; pls fix. Whether you go with the approach I suggested of having this return a status or no, pls update the names to reflect that one is for register and one is for login.

nashirj commented on lib/screens/UsersProfileScreen.dart:
```diff
@@ -0,0 +1,1576 @@
+import 'package:flutter/material.dart';
+import 'dart:convert';
+import 'package:smart_chef/utils/APIutils.dart';
+import 'package:smart_chef/utils/colors.dart';
+import 'package:smart_chef/utils/userAPI.dart';
+import 'package:smart_chef/utils/authAPI.dart';
+import 'package:http/http.dart' as http;
```
Please read the section in the style guide about [import order](https://dart.dev/guides/language/effective-dart/style#ordering) and fix your imports here and elsewhere accordingly

nashirj commented on lib/screens/UsersProfileScreen.dart:
```diff
@@ -0,0 +1,1576 @@
+        backgroundColor: Colors.white,
+        leading: IconButton(
+          onPressed: () async{
+            //TODO Finish registration verification once API is done
```
I think I've said it elsewhere, but pls fix all this to include an issue number and have the correct format `// TODO(<num here>): descrip`

nashirj commented on lib/screens/UsersProfileScreen.dart:
```diff
@@ -0,0 +1,1576 @@
+            try {
+              final response = await Authentication.logout();
+              switch (response.statusCode) {
+                case 200:
+                  setState(() {
+                    errorMessage = 'Logout successful!';
+                  });
+                  await Future.delayed(Duration(seconds: 1));
+                  user.clear();
+
+                  Navigator.pushNamedAndRemoveUntil(context,
+                      '/startup', ((Route<dynamic> route) => false));
+                  break;
+                case 400:
+                  print("Could not logout");
+                  break;
+                case 401:
+                  print('Access Token missing');
+                  break;
+                default:
+                  print('Something went wrong!');
+                  break;
```
Please clean this up similar to what you have in your other screen

nashirj commented on lib/screens/UsersProfileScreen.dart:
```diff
@@ -0,0 +1,1576 @@
+
+              user.clear();
+
+              try {
+                final response = await User.deleteUser();
+                switch (response.statusCode) {
+                  case 200:
+                    user.accessToken = '';
+                    user.refreshToken = '';
+
+                    Navigator.pushNamedAndRemoveUntil(context,
+                        '/startup', ((Route<dynamic> route) => false));
+                    break;
+                  case 400:
+                    print("Incorrect request format");
+                    break;
+                  case 401:
+                    print('Access Token missing');
+                    break;
+                  case 404:
+                    errorMessage = "User wasn't found";
+                    break;
+                  default:
+                    print('Something went wrong!');
+                    break;
+                }
```
Pls extract

nashirj commented on lib/screens/UsersProfileScreen.dart:
```diff
@@ -0,0 +1,1576 @@
+                    break;
+                }
+
+                //jsonObject = json.decode(ret.body);
```
Remove if not needed

nashirj commented on lib/screens/UsersProfileScreen.dart:
```diff
@@ -0,0 +1,1576 @@
+                          ),
+                          Container(
+                            width: 180,
+                            //padding: EdgeInsets.all(15),
```
Pls remove if not needed, ditto for all others

nashirj commented on lib/screens/UsersProfileScreen.dart:
```diff
@@ -0,0 +1,1576 @@
+                              },
+                            ),
+                          ),
+                          Container(
```
When(/if) you have time, can you look into reusable components? There is a decent amount of duplication with small changes; I imagine you could parameterize stuff so that you can include a child container here and pass in whatever specific stuff (e.g. `hintText`). That would cut down on duplication a lot here and elsewhere - obviously it's not a priority, but if you have time and interest I think it would make the codebase cleaner and be a good learning experience for you (and me, living vicariously through you :laughing: )

smaliszewski15 replied:
I have definitely thought about doing this, but was gonna leave it until most of the important stuff was done first. I can make it a sooner rather than later thing though if desired.

nashirj commented on lib/screens/UsersProfileScreen.dart:
```diff
@@ -0,0 +1,1576 @@
+                              textAlign: TextAlign.left,
+                              textInputAction: TextInputAction.done,
+                              onChanged: (password) {
+                                if (password.isEmpty) {
+                                  setState(() => unfilledConfirmPassword = true);
+                                }
+                                else {
+                                  if (_password.value.text.isEmpty) {
+                                    errorMessage = 'Passwords must match!';
+                                  }
+                                  else if (passwordsMatch()) {
+                                    errorMessage = '';
+                                    unfilledConfirmPassword = false;
+                                  }
+                                  setState(() {});
+                                }
```
Elsewhere, you have a validation function for this - I think you can do the same here

nashirj commented on lib/screens/UsersProfileScreen.dart:
```diff
@@ -0,0 +1,1576 @@
+    );
+  }
+
+  bool areFieldsValid() {
```
Pls make name more specific same comment as above

nashirj commented on lib/screens/UsersProfileScreen.dart:
```diff
@@ -0,0 +1,1576 @@
+    }
+    if (!isEmail(_email.value.text)) {
+      toReturn = false;
+      errorMessage = 'Email must be in valid form';
```
If you have `errorMessage` for one, consider having for all

nashirj commented on lib/screens/UsersProfileScreen.dart:
```diff
@@ -0,0 +1,1576 @@
+                                    String firstName = _firstName.value.text.trim();
+                                    String lastName = _lastName.value.text.trim();
+                                    String username = _email.value.text.trim();
+                                    String changes = '{"firstname": "$firstName","lastname": "$lastName","lastSeen": ${user
+                                        .lastSeen}"username": "$username","password": "${user
```
I think `lastSeen` is computed server side, I don't think you need to worry about including that

nashirj commented on lib/screens/UsersProfileScreen.dart:
```diff
@@ -0,0 +1,1576 @@
+    return true;
+  }
+
+  void clearFields() {
```
It doesn't look like this is used in this file, should it be? If not, pls delete

smaliszewski15 replied:
I think I just forgot to include it in the code.

nashirj commented on lib/screens/UsersProfileScreen.dart:
```diff
@@ -0,0 +1,1576 @@
+    _password.clear();
+  }
+
+  Future<bool> status(http.Response res) async {
```
Pls make name more descriptive. Something like `validateUpdateUserResponse`. Ditto elsewhere where you validate a response

nashirj commented on lib/screens/UsersProfileScreen.dart:
```diff
@@ -0,0 +1,1576 @@
+          errorMessage = 'Access token missing';
+          return false;
+        case 403:
+          final changeToken = await Authentication.refreshToken();
```
`refreshTokenResponse`

nashirj commented on lib/screens/UsersProfileScreen.dart:
```diff
@@ -0,0 +1,1576 @@
+            user.refreshToken = tokens['refreshToken'];
+          }
+          else {
+            print('something went wrong!');
```
If you end up here, you should try logging the user in again using the saved info and if that fails, redirect to startup

nashirj commented on lib/screens/UsersProfileScreen.dart:
```diff
@@ -0,0 +1,1576 @@
+                                  : globalDecoration.copyWith(
+                                  hintText: 'Confirm Password'),
+                              onChanged: (password) {
+                                if (password.isEmpty) {
+                                  setState(() =>
+                                  unfilledConfirmPassword = true);
+                                }
+                                else {
+                                  if (_newPassword.value.text.isEmpty |
+                                  (_newPassword.value.text != password)) {
+                                    errorMessage = 'Passwords must match!';
+                                  }
+                                  else {
+                                    errorMessage = '';
+                                    setState(() => unfilledConfirmPassword = false);
+                                  }
+                                }
```
Pls extract

nashirj commented on lib/screens/UsersProfileScreen.dart:
```diff
@@ -0,0 +1,1576 @@
+                            padding: EdgeInsets.all(15),
+                            margin: EdgeInsets.only(top: 170),
+                            child: ElevatedButton(
+                              onPressed: () async {
```
Pls clean this up like you did with your other views

nashirj commented on lib/screens/UsersProfileScreen.dart:
```diff
@@ -0,0 +1,1576 @@
+                                            'Access token missing';
+                                            break;
+                                          case 403:
+                                            final changeToken = await Authentication.refreshToken();
+                                            switch (changeToken.statusCode) {
+                                              case 200:
+                                                var tokens = json.decode(
+                                                    changeToken.body);
+                                                user.accessToken =
+                                                tokens['accessToken'];
+                                                user.refreshToken =
+                                                tokens['refreshToken'];
+                                                break;
+                                              case 400:
+                                                errorMessage =
+                                                'Could not refresh tokens';
+                                                break;
+                                              case 401:
+                                                errorMessage = 'Token missing';
+                                                break;
+                                              case 404:
+                                                errorMessage = 'User not found';
+                                                break;
+                                            }
```
Since you do this in a couple places, you can have a dedicated function somewhere for refreshing the token/relogging in the user if the token expired; you're prob going to need that logic for every endpoint that requires the token. This can go in a helper util file.

nashirj reviewed (commented):
Really nice changes, your code is way improved over the first draft! Left you some more comments. Also, please update all your variable/class/etc names according to the [Dart style guide](https://dart.dev/guides/language/effective-dart/style)

smaliszewski15 committed:
Ingredients page merge

smaliszewski15 committed:
Made most requested changes to all files

smaliszewski15 committed:
Refactored errors in invalid text fields to avoid repetition
Fixed a small typo

smaliszewski15 committed:
Refactored errors in invalid text fields to avoid repetition
Fixed a small typo

smaliszewski15 committed:
Added issue numbers to corresponding TODO's

smaliszewski15 committed:
Added issue numbers to corresponding TODO's

smaliszewski15 committed:
More small changes

smaliszewski15 committed:
More small changes

smaliszewski15 committed:
Hotfixes

smaliszewski15 committed:
Hotfixes

smaliszewski15 committed:
Fixed the verification process
Added way for users on log in who haven't verified their account yet to verify their account
Refactored a few things
Fixed more typos/bugs

smaliszewski15 committed:
Fixed the verification process
Added way for users on log in who haven't verified their account yet to verify their account
Refactored a few things
Fixed more typos/bugs

smaliszewski15 committed:
Hotfixes

smaliszewski15 committed:
Hotfixes

nashirj commented on lib/routes/routes.dart:
```diff
@@ -7,11 +7,15 @@ import 'package:smart_chef/screens/UsersProfileScreen.dart';
-  static const String ingredientScreen = '/food';
+
+  static const String ingredientsScreen = '/food';
+  static const String ingredientScreen = '/food/food';
```
This could be easy to typo/get mixed up - can it be `ingredientsScreen` and `individualIngredientScreen` instead?

nashirj commented on lib/screens/RecipeScreen.dart:
```diff
@@ -43,6 +47,7 @@ class _RecipePageState extends State<RecipePage> {
           centerTitle: true,
           backgroundColor: Colors.white,
           leading: IconButton(
+            // TODO(): Make search functionality
```
Please create an issue and add the corresponding issue number, ditto elsewhere

nashirj commented on lib/screens/RecipeScreen.dart:
```diff
@@ -93,13 +100,12 @@ class _RecipePageState extends State<RecipePage> {
                           'Ingredients',
-                          style:
-                          TextStyle(fontSize: 12, color: bottomRowIcon),
+                          style: TextStyle(fontSize: 12, color: bottomRowIcon),
```
Can you please make a variable for this font size if it's used in a lot of places?

nashirj commented on lib/utils/userData.dart:
```diff
@@ -1,16 +1,50 @@
+    firstName = json['firstName'];
+    lastName = json['lastName'];
+    username = json['username'];
+    email = json['email'];
```
can you add `this` to these just to make it extra clear what's happening? same below

nashirj commented on lib/utils/globals.dart:
```diff
@@ -0,0 +1,49 @@
+import 'package:flutter/material.dart';
+import 'package:smart_chef/utils/colors.dart';
+
+const double bottomIconSize = 55;
```
Can you add variables for all the stuff you have hardcoded numbers for? e.g. `topIconSize`, and text fonts and stuff?

nashirj commented on lib/utils/authAPI.dart:
```diff
@@ -4,68 +4,99 @@ import 'dart:convert';
 import 'package:smart_chef/utils/APIutils.dart';
```
This code looks so clean! Really nice work :)

nashirj commented on lib/utils/APIutils.dart:
```diff
@@ -1,12 +1,85 @@
   return emailValidation.hasMatch(email);
 }
+
+//Function that will go through the process of attempting to refresh the token
```
Please have a single space between `//` and start of comment

nashirj commented on lib/utils/APIutils.dart:
```diff
@@ -1,12 +1,85 @@
+  }
+}
+
+Future<bool> relogUser() async {
```
Maybe `reauthenticateUser`?

nashirj commented on lib/screens/StartupScreen.dart:
```diff
@@ -483,7 +520,7 @@ class _SignInPageState extends State<SignInPage> {
                         topMessage = 'Forgot Your\nPassword?';
                         setState(() {
-                          state = 1;
+                          //state = 1;
```
Pls add a TODO

nashirj commented on lib/screens/StartupScreen.dart:
```diff
@@ -250,7 +244,8 @@ class _SignInPageState extends State<SignInPage> {
                     ),
                   ),
-                  detectState(),
+                  //detectState(),
```
Pls delete if not needed or add TODO

nashirj commented on lib/screens/StartupScreen.dart:
```diff
@@ -327,25 +322,17 @@ class _SignInPageState extends State<SignInPage> {
-                                setState(() => unfilledEmail = false);
+                            controller: _username,
+                            decoration: unfilledUsername
+                                ? invalidTextField.copyWith(
```
Nice :) much cleaner!

nashirj commented on lib/screens/StartupScreen.dart:
```diff
@@ -414,38 +402,87 @@ class _SignInPageState extends State<SignInPage> {
-                                    );
+                                  final ret =
+                                      await Authentication.login(payload);
+                                  print(ret.statusCode);
```
Can remove this printline

nashirj commented on lib/screens/StartupScreen.dart:
```diff
@@ -498,188 +535,199 @@ class _SignInPageState extends State<SignInPage> {
-      ],
-    );
-  }
+  //TODO(): Add support for reseting password
```
I don't mind leaving this in, but an alternative to having all this commented out code in the main branch would be to make a new branch based off this one and put the code in there until you need it.

nashirj commented on lib/screens/StartupScreen.dart:
```diff
@@ -1009,292 +1421,96 @@ class _RegisterPageState extends State<RegisterPage> {
+                          try {
+                            final res =
+                                await Authentication.verifyCode(payload);
+                            print(res.statusCode);
```
Can remove print

nashirj commented on lib/screens/StartupScreen.dart:
```diff
@@ -1009,292 +1421,96 @@ class _RegisterPageState extends State<RegisterPage> {
+                                  'username': user.username,
+                                };
+                                final ret = await Authentication.sendCode(name);
+                                //print(ret.statusCode);
```
Pls remove

nashirj commented on lib/screens/UsersProfileScreen.dart:
```diff
@@ -1349,93 +1275,54 @@ class _EditPasswordPageState extends State<EditPasswordPage> {
+                                  try {
+                                    final res =
+                                        await User.updateUser(changes);
+                                    if (res.statusCode == 200) {
```
I might be wrong, but I think 1290-1311 can be simplified - if this doesn't work, np, but you should be able to take advantage of `Navigator.pop(context)`
```dart
if (res.statusCode == 200) {
  user.password = _newPassword.value.text.trim();
  errorMessage == 'Successfully updated your password!';
  await Future.delayed(Duration(seconds: 1));
  clearFields();
  Navigator.pop(context);
}
// Note (don't need to include this in code): we only get to this code if the above
// is not true, since we leave the page after successful execution of the above
errorDialog(context);
```

nashirj reviewed (approved):
Really nice work! Thanks for incorporating all my suggestions

smaliszewski15 committed:
Made requested changes

smaliszewski15 committed:
Merge remote-tracking branch 'origin/attaching-api' into attaching-api

# Conflicts:
#	lib/routes/routes.dart
#	lib/screens/RecipeScreen.dart
#	lib/screens/ShoppingCartScreen.dart
#	lib/screens/StartupScreen.dart
#	lib/screens/UsersProfileScreen.dart
#	lib/utils/APIutils.dart
#	lib/utils/colors.dart
#	lib/utils/globals.dart
#	lib/utils/userData.dart
